docs/books/valhalla-mjolnir/src/02_osm_data_processing.md

```diff
@@ -0,0 +1,282 @@
+class OSMWay {
+public:
+  OSMWay() = default;
+  
+  // Set way attributes based on tags
+  void set_node_ids(const std::vector<uint64_t>& node_ids);
+  void set_name(const std::string& name);
+  void set_road_class(const RoadClass roadclass);
+  // ... many more setters ...
+  
+  // Get way attributes
+  const std::vector<uint64_t>& node_ids() const;
+  const std::string& name() const;
+  RoadClass road_class() const;
+  // ... many more getters ...
+  
+private:
+  std::vector<uint64_t> node_ids_;
+  std::string name_;
+  std::string name_en_;
+  std::string alt_name_;
+  std::string official_name_;
+  // ... many more fields ...
+};
+```
+
+### Relation Processing
+
+Finally, relations are processed to capture complex structures like turn restrictions:
+
+```cpp
+// From src/mjolnir/pbfgraphparser.cc
+void PBFGraphParser::ParseRelations(const OSMPBF::PrimitiveBlock& primblock) {
+  // For each primitive group
+  for (int i = 0; i < primblock.primitivegroup_size(); ++i) {
+    const OSMPBF::PrimitiveGroup& pg = primblock.primitivegroup(i);
+    
+    // Relations
+    for (int j = 0; j < pg.relations_size(); ++j) {
+      const OSMPBF::Relation& relation = pg.relations(j);
+      
+      // Get tags
+      std::vector<Tag> tags = GetTagsFromRelation(primblock, relation);
+      
+      // Check if this is a restriction relation
+      if (IsRestriction(tags)) {
+        // Process restriction
+        // ...
+      }
+    }
+  }
+}
+```
+
+Relations are converted to Valhalla's internal `OSMRestriction` structure:
+
+```cpp
+// From valhalla/mjolnir/osmrestriction.h
+struct OSMRestriction {
+  OSMRestriction() = default;
+  
+  uint64_t from_way_id;
+  uint64_t to_way_id;
+  uint64_t via_node_id;
+  uint64_t via_way_id;
+  
+  RestrictionType type;
+  uint32_t day_on;
+  uint32_t day_off;
+  uint32_t hour_on;
+  uint32_t hour_off;
+};
+```
+
+## Tag Processing and Transformation
+
+OSM tags (key-value pairs) contain information about road properties. Mjolnir processes these tags to extract:
+
+- Road classification (highway, residential, footway, etc.)
+- Access restrictions (vehicle types allowed)
+- Speed limits
+- Direction of travel (one-way, bidirectional)
+- Surface type and quality
+- And many more attributes
+
+Mjolnir includes a tag transformation system that can be customized using Lua scripts:
+
+```cpp
+// From valhalla/mjolnir/luatagtransform.h
+class LuaTagTransform {
+public:
+  LuaTagTransform(const std::string& lua_script);
+  ~LuaTagTransform();
+  
+  // Transform tags for ways
+  bool Transform(OSMWay& way, const Tags& tags);
+  
+  // Transform tags for nodes
+  bool Transform(OSMNode& node, const Tags& tags);
+  
+private:
+  lua_State* lua_state_;
+};
+```
+
+This allows for flexible customization of how OSM tags are interpreted without modifying the C++ code. The Lua scripts transform the schemaless OSM tags into a structured set of values that can be efficiently processed by the C++ code.
```

Comment:
sadly there is still a bunch of c++ side wrangling. so much that ive been often tempted to just port the lua transformations into c++ and forgo lua all together. not enough time on my hands though!

Reply:
This is actually a great point. I see a need of defining profiles for routes on Overture data. But I have not formed an opinion on how that should look like. 

The ways I see we can do it:
1. Scripting solutions like Lua.
2. Plugins in the native language, loaded while building tiles or serving routes.